### ISSUE TITLE
Add in rancher-compose support for missing rancher-compose/docker-compose fields

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#6280]
As docker-compose/rancher-compose has added supported fields, we are missing the support within Rancher. 

After Rancher-compose has added support, there are additional areas that need to be updated
1. Adding the field to the container API
2. Adding the field to the exported docker-compose.yml of the stack
   **3. Adding the field to be read/imported by rancher-compose - This ticket is to track being able to import these fields using docker-compose.yml** 
3. UI (Once Cattle is complete, we can track the UI in a separate issue)

Known Fields that are missing.

| Field | Description |
| --- | --- |
| \- [ ] --add-host=[] | Add a custom host-to-IP mapping (host:ip) |
| \- [ ]   --blkio-weight | Block IO (relative weight), between 10 and 1000 |
| \- [ ]   --blkio-weight-device=[] | Block IO weight (relative device weight) |
| \- [ ]  --cpu-shares | CPU shares (relative weight) |
| \- [ ]  --cgroup-parent | Optional parent cgroup for the container |
| \- [ ]  --cpu-period | Limit CPU CFS (Completely Fair Scheduler) period |
| \- [ ]  --cpu-quota | Limit CPU CFS (Completely Fair Scheduler) quota |
| \- [ ]  --cpuset-cpus | CPUs in which to allow execution (0-3, 0,1) |
| \- [ ]  --cpuset-mems | MEMs in which to allow execution (0-3, 0,1) |
| \- [ ]  --device-read-bps=[] | Limit read rate (bytes per second) from a device |
| \- [ ]  --device-read-iops=[] | Limit read rate (IO per second) from a device |
| \- [ ]  --device-write-bps=[] | Limit write rate (bytes per second) to a device |
| \- [ ]  --device-write-iops=[] | Limit write rate (IO per second) to a device |
| \- [ ]  --dns-opt=[] | Set DNS options |
| \- [ ]  --dns-search=[] | Set custom DNS search domains |
| \- [ ]  --group-add=[] | Add additional groups to join |
| \- [ ]  --ip | Container IPv4 address (e.g. 172.30.100.104) |
| \- [ ]  --ip6 | Container IPv6 address (e.g. 2001:db8::33) |
| \- [ ]  --ipc | IPC namespace to use |
| \- [ ]  --isolation | Container isolation level |
| \- [ ]  --kernel-memory | Kernel memory limit |
| \- [ ]  --memory-reservation | Memory soft limit |
| \- [ ]  --memory-swap | Swap limit equal to memory plus swap: '-1' to enable unlimited swap |
| \- [ ]  --memory-swappiness=-1 | Tune container memory swappiness (0 to 100) |
| \- [ ]  --net-alias=[] | Add network-scoped alias for the container |
| \- [ ]  --oom-kill-disable | Disable OOM Killer |
| \- [ ]  --oom-score-adj | Tune host's OOM preferences (-1000 to 1000) |
| \- [ ]  --read-only | Mount the container's root filesystem as read only |
| \- [ ]  --security-opt=[] | Security Options |
| \- [ ]  --shm-size | Size of /dev/shm, default value is 64MB |
| \- [ ]  --stop-signal=SIGTERM | Signal to stop a container, SIGTERM by default |
| \- [ ]  --sysctl value | Sysctl options (default map[]) |
| \- [ ]  --tmpfs=[] | Mount a tmpfs directory |
| \- [ ]  --ulimit=[] | Ulimit options |
| \- [ ]  --uts | UTS namespace to usea |

**Comment by deniseschannon** [comment]
The original task has been split into 4 issues:

Cattle Changes: #4708
Rancher-compose: #6280
Exporting yml: #6281
UI: #6282

**Comment by joshwget** [comment]
Some fields are not covered by rancher/rancher-compose#206.

Unknown what key Docker Compose will use for these attributes:
- `cpuset-mems`
- `kernel-memory`
- `sysctl`

Fields dependent on #5715:
- `ip`?
- `ip6`?
- `net-alias`

**Comment by soumyalj** [comment]
Tested with v1.2.0-pre4-rc7. All the keys mentioned above and the ones supported are working. 
Created a stack using the yml file below and the stack creation was successful:
```
version: '2'
services:
  test:
    device_write_bps:
      /dev/null: 200
    device_write_iops:
      /dev/vda: 3000
    cpu_quota: 20000
    read_only: true
    memswap_limit: 2000000000
    oom_score_adj: 100
    cpu_shares: 400
    shm_size: 1024
    cgroup_parent: abcd
    mem_limit: 1000000000
    security_opt:
    - label:user:USER
    - label:role:ROLE
    tmpfs:
    - /run
    device_read_bps:
      /dev/null: 4000
    group_add:
    - root
    blkio_weight: 1000
    stop_signal: SIGTERM
    dns_opt:
    - abc
    extra_hosts:
    - h1:162.242.195.82
    - h2:162.241.195.82
    cpu_period: 10000
    cpuset: '0'
    image: ibuildthecloud/helloworld:latest
    mem_reservation: 5000000
    device_read_iops:
      /dev/vda: 2000
    dns_search:
    - abc
    mem_swappiness: 100
    oom_kill_disable: true
    ipc: host
    isolation: default
```
Issue related to isolation key is https://github.com/rancher/rancher/issues/6587

**Comment by kimgust** [comment]
Fwiw, if I set `shm_size: 1024` in docker-compose.yml per the [example above](https://github.com/rancher/rancher/issues/6280#issuecomment-259833332) I get a /dev/shm size of 4K. If I set it to a string format like 512M per the [docs](https://docs.docker.com/compose/compose-file/#/cpushares-cpuquota-cpuset-domainname-hostname-ipc-macaddress-memlimit-memswaplimit-oomscoreadj-privileged-readonly-restart-shmsize-stdinopen-tty-user-workingdir), I get:
```
ERRO[0000] Could not parse config for project server : strconv.ParseInt: parsing "512M": invalid syntax 
FATA[0000] Failed to read project: strconv.ParseInt: parsing "512M": invalid syntax 
```
If I set it to `shm_size: 536870912` I get the expected 512M value.

![image](https://cloud.githubusercontent.com/assets/299082/22045404/eedcafaa-dcce-11e6-8ed0-8ab84f7d026e.png)

```
$ docker --version
Docker version 1.12.1, build 23cf638
```

**Comment by joshwget** [comment]
@kimgust Support for this was added to libcompose pretty recently (https://github.com/docker/libcompose/pull/427) and it should make its way into Rancher soon.

**Comment by eesprit** [comment]
Any news on having the sysctls supported ?

**Comment by joshwget** [comment]
@eesprit `sysctls` key will be added in 1.5.4.
